File: LeetCode295-Find Median from a Data Stream/median.py
from heapq import *
class MedianFinder(object):

    def __init__(self):
        self.minHeap = []
        self.maxHeap = []

    def addNum(self, num):
        """
        :type num: int
        :rtype: None
        """
        if not self.maxHeap or -self.maxHeap[0] >= num :
            heappush(self.maxHeap, -num)
        else:
            heappush(self.minHeap, num)
        
        self.balance()
    
    def balance(self):
        if len(self.maxHeap) > len(self.minHeap) + 1:
            heappush(self.minHeap, -heappop(self.maxHeap))
        elif len(self.maxHeap) < len(self.minHeap):
            heappush(self.maxHeap, -heappop(self.minHeap))

    def findMedian(self):
        """
        :rtype: float
        """
        if len(self.maxHeap) > len(self.minHeap):
            return -self.maxHeap[0]/1.0
        else:
            return -self.maxHeap[0]/2.0 + self.minHeap[0]/2.0

if __name__ == "__main__":
    dataStream = [2, 3, 7, 9]
    medianFinder = MedianFinder()

    for num in dataStream:
        print("add number", num)
        medianFinder.addNum(num)
        print("max heap", medianFinder.maxHeap)
        print("min heap", medianFinder.minHeap)
        print("median", medianFinder.findMedian())


"""
output:
('add number', 2)
('max heap', [-2])
('min heap', [])
('median', 2.0)

('add number', 3)
('max heap', [-2])
('min heap', [3])
('median', 2.5)

('add number', 7)
('max heap', [-3, -2])
('min heap', [7])
('median', 3.0)

('add number', 9)
('max heap', [-3, -2])
('min heap', [7, 9])
('median', 5.0)
"""

File: LeetCode295-Find Median from a Data Stream/median.go
package main

import (
	"container/heap"
	"fmt"
)

type MedianFinder struct {
	minHeap *MinHeap
	maxHeap *MaxHeap
}

func Constructor() MedianFinder {
	return MedianFinder{
		minHeap: &MinHeap{},
		maxHeap: &MaxHeap{},
	}
}

func (this *MedianFinder) AddNum(num int) {
	if this.maxHeap.Len() == 0 || this.maxHeap.Top() >= num {
		heap.Push(this.maxHeap, num)
	} else {
		heap.Push(this.minHeap, num)
	}

	this.balance()
}

func (this *MedianFinder) balance() {
	if this.maxHeap.Len() > this.minHeap.Len()+1 {
		heap.Push(this.minHeap, heap.Pop(this.maxHeap).(int))
	} else if this.maxHeap.Len() < this.minHeap.Len() {
		heap.Push(this.maxHeap, heap.Pop(this.minHeap).(int))
	}
}

func (this *MedianFinder) FindMedian() float64 {
	if this.maxHeap.Len() > this.minHeap.Len() {
		return float64(this.maxHeap.Top())
	} else {
		return float64(float64(this.maxHeap.Top()+this.minHeap.Top()) / 2.0)
	}
}

type MaxHeap []int

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}
func (h MaxHeap) Top() int { return h[0] }

type MinHeap []int

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}
func (h MinHeap) Top() int { return h[0] }

func main() {
	dataSteam := []int{2, 3, 7, 9}
	medianFinder := Constructor()

	for _, num := range dataSteam {
		fmt.Printf("add number: %d\n", num)
		medianFinder.AddNum(num)
		fmt.Println("max heap", medianFinder.maxHeap)
		fmt.Println("min heap", medianFinder.minHeap)
		fmt.Println("median", medianFinder.FindMedian())
	}
}

/*
output:
add number: 2
max heap &[2]
min heap &[]
median 2

add number: 3
max heap &[2]
min heap &[3]
median 2.5

add number: 7
max heap &[3 2]
min heap &[7]
median 3

add number: 9
max heap &[3 2]
min heap &[7 9]
median 5
*/
